src/Geometry/Primitives/Zernike.jl

```diff
@@ -167,81 +163,51 @@ struct ZernikeSurface{T,N,P,Q} <: ParametricSurface{T,N}
             end
         end
         P = length(zcs)
-        acs = []
-        if aspherics !== nothing
-            for (i, k) in aspherics
-                if abs(k) > zero(T)
-                    @assert i % 2 == 0 "Only even aspheric terms are supported"
-                    push!(acs, (i ÷ 2, k))
-                end
-            end
-        end
-        Q = length(acs)
-        new{T,3,P,Q}(semidiameter, SVector{P,Tuple{Int,Int,T}}(zcs), SVector{Q,Tuple{Int,T}}(acs), 1 / radius, conic, normradius, Cylinder(semidiameter, interface = opaqueinterface(T))) # TODO!! incorrect interface on cylinder
+        M=asphericType(asp)
+        new{T,3,P,Q,M}(asp, SVector{P,Tuple{Int,Int,T}}(zcs), Cylinder(semidiameter, interface = opaqueinterface(T))) # TODO!! incorrect interface on cylinder
     end
 
 end
 export ZernikeSurface
 
 uvrange(::Type{ZernikeSurface{T,N,P,Q}}) where {T<:Real,N,P,Q} = ((zero(T), one(T)), (-T(π), T(π))) # ρ and ϕ
 
-semidiameter(z::ZernikeSurface{T}) where {T<:Real} = z.semidiameter
+semidiameter(z::ZernikeSurface{T}) where {T<:Real} = z.asp.semidiameter
 halfsizeu(z::ZernikeSurface{T}) where {T<:Real} = semidiameter(z)
 halfsizev(z::ZernikeSurface{T}) where {T<:Real} = semidiameter(z)
 
 boundingobj(z::ZernikeSurface{T}) where {T<:Real} = z.boundingcylinder
 
 function point(z::ZernikeSurface{T,3,P,Q}, ρ::T, ϕ::T)::SVector{3,T} where {T<:Real,P,Q}
-    rad = z.semidiameter
-    r = ρ * rad
-    r2 = r^2
-    t = one(T) - z.curvature^2 * (z.conic + one(T)) * r^2
-    if t < zero(T)
-        return SVector{3,T}(NaN, NaN, NaN)
-    end
-    h = z.curvature * r2 / (one(T) + sqrt(t))
-    # sum aspheric
-    @inbounds @simd for i in 1:Q
-        (m, k) = z.aspherics[i]
-        h += k * r2^m
-    end
+    pnt = point(z.asp, ρ, ϕ)
+
     # sum zernike
-    u = r / z.normradius
+    rad=semidiameter(z.asp)
```

Comment:
```suggestion
    rad = semidiameter(z.asp)
```

Also L168, L197 and L199
And L85, 89 and 93